nlmod/dims/layers.py

```diff
@@ -2003,3 +2005,116 @@ def get_isosurface(da, z, value, input_core_dims=None, exclude_dims=None, **kwar
         dask="forbidden",
         **kwargs,
     )
+
+
+def get_modellayers_screens(ds, screen_top, screen_bottom, xy=None, icell2d=None):
+    """Get the modellayer of a well based on a screen_top, screen_bottom and xy
+    coordinates. 
+
+    Parameters
+    ----------
+    ds : xarray.Dataset
+        xarray Dataset with with top and bottoms, can be structured or vertex.
+    screen_top : np.ndarray of shape(nobs)
+        collection of screen top values).
+    screen_bottom : np.ndarray of shape(nobs)
+        collection of screen bottom values. Should have the same length as screen_top
+        and xy.
+    xy : np.ndarray of shape(nobs, 3), optional
+        list of x,y coordinates
+    icell2d : np.ndarray of shape(nobs), optional
+        To speed up the process for vertex grids a list of icell2d indices can be 
+        given instead of a list of xy coordinates.
+
+    Returns
+    -------
+    list of floats
+        zero-based indices of modellayers
+        nan if screen above or below model boundaries
+        if screen spans multiple layers, choose layer with most screen length.
+    """
+    if grid.is_vertex(ds):
+        if icell2d is None:
+            gi = flopy.utils.GridIntersect(grid.modelgrid_from_ds(ds), method="vertex")
+            icell2d = [grid.get_icell2d_from_xy(x, y, ds, gi=gi) for x,y in xy]
+        # make dataset of observations
+        ds_obs = ds.sel(icell2d=icell2d)
+        ds_obs['screen_top'] = (('icell2d'), screen_top)
+        ds_obs['screen_bot'] = (('icell2d'), screen_bottom)
+        dimname = 'icell2d'
+    elif grid.is_structured(ds):
+        # make dataset of observations
+        dimname = 'no_obs'
+        x = xr.DataArray(np.asarray(xy)[:,0], dims=dimname)
+        y = xr.DataArray(np.asarray(xy)[:,1], dims=dimname)
+        ds_obs = ds.sel(x=x,y=y, method='nearest')
+        ds_obs['screen_top'] = ((dimname), screen_top)
+        ds_obs['screen_bot'] = ((dimname), screen_bottom)
+
+    modellayers = _get_modellayers_dsobs(ds_obs, dimname=dimname)
+    return modellayers
+
+def _get_modellayers_dsobs(ds_obs, dimname='no_obs'):
+    """Get modellayers from a dataset of observation point data
+
+    Parameters
+    ----------
+    ds_obs : xr.Dataset
+        typically a subset of a model dataset with only data for observations.
+    dimname : str, optional
+        name of the observation dimension. Usually 'icell2d' or 'no_obs', 
```

Comment:
maybe indicate that one is used for unstructured models, and the other for structured?